File: benchmarks/tasks/type_definition/type_definition.py
# SPDX-License-Identifier: MIT-0

class Nil:
    pass

class Cons:
    def __init__(self, head, tail):
        self.head = head
        self.tail = tail

def length(xs):
    if isinstance(xs, Nil):
        return 0
    return 1 + length(xs.tail)

def append(xs, x):
    if isinstance(xs, Nil):
        return Cons(x, Nil())
    return Cons(xs.head, append(xs.tail, x))

xs = Cons(1, Cons(2, Cons(3, Nil())))
ys = append(xs, 4)
print(length(ys))

File: benchmarks/tasks/type_definition/type_definition.js
// SPDX-License-Identifier: MIT-0

function nil() { return { type: "nil" }; }
function cons(head, tail) { return { type: "cons", head, tail }; }

function length(xs) {
    if (xs.type === "nil") return 0;
    return 1 + length(xs.tail);
}

function append(xs, x) {
    if (xs.type === "nil") return cons(x, nil());
    return cons(xs.head, append(xs.tail, x));
}

const xs = cons(1, cons(2, cons(3, nil())));
const ys = append(xs, 4);
console.log(length(ys));
